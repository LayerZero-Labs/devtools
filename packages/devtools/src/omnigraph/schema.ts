import { EndpointId } from '@layerzerolabs/lz-definitions'
import { z } from 'zod'
import type { OmniPoint, OmniNode, OmniVector, OmniEdge, OmniGraph } from './types'

export const UIntSchema = z.coerce.number().nonnegative().int()

export const AddressSchema = z.string()

export const EndpointIdSchema: z.ZodSchema<EndpointId, z.ZodTypeDef, string | number> = z
    .nativeEnum(EndpointId)
    .pipe(z.number())

export const OmniPointSchema: z.ZodSchema<OmniPoint, z.ZodTypeDef, unknown> = z.object({
    address: AddressSchema,
    eid: EndpointIdSchema,
})

export const OmniVectorSchema: z.ZodSchema<OmniVector, z.ZodTypeDef, unknown> = z.object({
    from: OmniPointSchema,
    to: OmniPointSchema,
})

export const EmptyOmniNodeSchema = z.object({
    point: OmniPointSchema,
    config: z.unknown().optional(),
})

export const EmptyOmniEdgeSchema = z.object({
    vector: OmniVectorSchema,
    config: z.unknown().optional(),
})

/**
 * Helper assertion utility for `OmniPoint` instances
 *
 * @param {unknown} value
 * @returns {boolean} `true` if the value is an `OmniPoint`, `false` otherwise
 */
export const isOmniPoint = (value: unknown): value is OmniPoint => OmniPointSchema.safeParse(value).success

/**
 * Helper assertion utility that checks whether an `OmniGraph`
 * has at least any contracts or connections defined
 *
 * @param {OmniGraph} graph
 * @returns {boolean}
 */
export const isOmniGraphEmpty = ({ contracts, connections }: OmniGraph): boolean =>
    contracts.length === 0 && connections.length === 0

/**
 * Factory for OmniNode schemas
 *
 * @param configSchema Schema of the config contained in the node
 *
 * @returns `z.ZodSchema<OmniNode<TConfig>>` schema for a node with the particular config type
 */
export const createOmniNodeSchema = <TConfig = unknown>(
    configSchema: z.ZodSchema<TConfig, z.ZodTypeDef, unknown>
): z.ZodSchema<OmniNode<TConfig>, z.ZodTypeDef> =>
    EmptyOmniNodeSchema.extend({
        config: configSchema,
    }) as z.ZodSchema<OmniNode<TConfig>, z.ZodTypeDef>

/**
 * Factory for OmniEdge schemas
 *
 * @param {z.ZodSchema<TConfig>} configSchema Schema of the config contained in the edge
 *
 * @returns {z.ZodSchema<OmniEdge<TConfig>>} Schema for an edge with the particular config type
 */
export const createOmniEdgeSchema = <TConfig = unknown>(
    configSchema: z.ZodSchema<TConfig, z.ZodTypeDef, unknown>
): z.ZodSchema<OmniEdge<TConfig>, z.ZodTypeDef> =>
    EmptyOmniEdgeSchema.extend({
        config: configSchema,
    }) as z.ZodSchema<OmniEdge<TConfig>, z.ZodTypeDef>

/**
 * Factory for OmniGraph schemas
 *
 * @param {z.ZodSchema<OmniNode<TNodeConfig>, z.ZodTypeDef, unknown>} nodeSchema
 * @param {z.ZodSchema<OmniEdge<TEdgeConfig>, z.ZodTypeDef, unknown>} edgeSchema
 *
 * @returns {z.ZodSchema<OmniGraph<TNodeConfig, TEdgeConfig>, z.ZodTypeDef, unknown>}
 */
export const createOmniGraphSchema = <TNodeConfig = unknown, TEdgeConfig = unknown>(
    nodeSchema: z.ZodSchema<OmniNode<TNodeConfig>, z.ZodTypeDef, unknown>,
    edgeSchema: z.ZodSchema<OmniEdge<TEdgeConfig>, z.ZodTypeDef, unknown>
): z.ZodSchema<OmniGraph<TNodeConfig, TEdgeConfig>, z.ZodTypeDef, unknown> =>
    z.object({
        contracts: z.array(nodeSchema),
        connections: z.array(edgeSchema),
    })
