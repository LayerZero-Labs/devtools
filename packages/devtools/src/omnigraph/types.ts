import type { EndpointId } from '@layerzerolabs/lz-definitions'
import type { Factory, OmniAddress, WithOptionals } from '@/types'
import { OmniTransaction } from '..'

/**
 * OmniPoint identifies a point in omniverse, an omnichain universe.
 *
 * In layman terms this is a contract deployed on a particular network (represented by an endpoint).
 */
export type OmniPoint = WithEid<{
    address: OmniAddress
    contractName?: string | null
}>

/**
 * OmniVector identifies a vector in omniverse, an omnichain universe.
 *
 * In layman terms this is a directional connection between two contracts - two points of the omniverse
 */
export interface OmniVector {
    from: OmniPoint
    to: OmniPoint
}

/**
 * OmniError represents an arbitrary error that occurred on a particular point in omniverse.
 */
export interface OmniError<TError = unknown> {
    point: OmniPoint
    error: TError
}

/**
 * OmniNode represents a point in omniverse
 * with an additional piece of information attached
 */
export type OmniNode<TConfig = unknown> = WithOptionals<{
    point: OmniPoint
    config: TConfig
}>

/**
 * OmniEdge represents a connection between two points in omniverse
 * with an additional piece of information attached
 */
export type OmniEdge<TConfig = unknown> = WithOptionals<{
    vector: OmniVector
    config: TConfig
}>

/**
 * OmniGraph is a collection of nodes and edges of omniverse
 * that together represent an omnichain app a.k.a. OApp.
 *
 * For purposes of readability and to avoid overabstraction on the user end,
 * the names are set to be `contracts` rather than `nodes` and `connections` rather than `edges`
 */
export interface OmniGraph<TNodeConfig = unknown, TEdgeConfig = unknown> {
    contracts: OmniNode<TNodeConfig>[]
    connections: OmniEdge<TEdgeConfig>[]
}

/**
 * Helper type for inferring the type of element in the `contracts` array of an `OmniGraph`
 */
export type InferOmniNode<TOmniGraph extends OmniGraph> = TOmniGraph['contracts'][number]

/**
 * Helper type for inferring the type of element in the `connections` array of an `OmniGraph`
 */
export type InferOmniEdge<TOmniGraph extends OmniGraph> = TOmniGraph['connections'][number]

/**
 * Helper type that adds eid property to an underlying type
 */
export type WithEid<TValue> = TValue & { eid: EndpointId }

/**
 * Base interface for all SDKs
 */
export interface IOmniSDK {
    point: OmniPoint
}

/**
 * Base factory for all SDK factories
 */
export type OmniSDKFactory<TOmniSDK = IOmniSDK, TOmniPoint = OmniPoint> = Factory<[TOmniPoint], TOmniSDK>

/**
 * Helper type for configuration functions
 */
export type Configurator<TOmniGraph extends OmniGraph = OmniGraph, TOmniSDK = IOmniSDK> = (
    graph: TOmniGraph,
    createSdk: OmniSDKFactory<TOmniSDK>
) => Promise<OmniTransaction[]>
