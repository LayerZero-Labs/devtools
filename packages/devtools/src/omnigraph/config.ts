import type { Factory } from '@/types'
import type { Configurator, IOmniSDK, InferOmniEdge, InferOmniNode, OmniGraph, OmniSDKFactory } from './types'
import type { OmniTransaction } from '@/transactions/types'
import { flattenTransactions } from '@/transactions/utils'

export type CreateTransactionsFromOmniNodes<TOmniGraph extends OmniGraph = OmniGraph, TOmniSDK = IOmniSDK> = Factory<
    [InferOmniNode<TOmniGraph>, TOmniSDK, TOmniGraph, OmniSDKFactory<TOmniSDK>],
    OmniTransaction[] | OmniTransaction | null | undefined
>

export type CreateTransactionsFromOmniEdges<TOmniGraph extends OmniGraph = OmniGraph, TOmniSDK = IOmniSDK> = Factory<
    [InferOmniEdge<TOmniGraph>, TOmniSDK, TOmniGraph, OmniSDKFactory<TOmniSDK>],
    OmniTransaction[] | OmniTransaction | null | undefined
>

/**
 * Function that takes care of the boilerplate for node configuration functions.
 *
 * It will create an SDK for every node in the graph, then call `createTransactions`
 * with the node itself, the created SDK, the whole graph and the SDK factory.
 *
 * ```
 * const configureSomething = createConfigureNodes(async ({ config }: OmniNode<{ something: string }>, sdk) => {
 *   const something = await sdk.getSomething()
 *   if (something !== config.something) return []
 *
 *   return sdk.setSomething(config.something)
 * })
 * ```
 *
 * @template TOmniGraph
 * @template TOmniSDK
 * @param {CreateTransactionsFromOmniNodes<TOmniGraph, TOmniSDK>} createTransactions
 * @returns {Configurator<TOmniGraph, TOmniSDK>}
 */
export const createConfigureNodes =
    <TOmniGraph extends OmniGraph = OmniGraph, TOmniSDK = IOmniSDK>(
        createTransactions: CreateTransactionsFromOmniNodes<TOmniGraph, TOmniSDK>
    ): Configurator<TOmniGraph, TOmniSDK> =>
    async (graph, createSdk) =>
        flattenTransactions(
            await Promise.all(
                graph.contracts.map(async (node) => {
                    const sdk = await createSdk(node.point)

                    return await createTransactions(node, sdk, graph, createSdk)
                })
            )
        )

/**
 * Function that takes care of the boilerplate for edge configuration functions.
 *
 * It will create an SDK for every edge (using the `from` field) in the graph, then call `createTransactions`
 * with the edge itself, the created SDK, the whole graph and the SDK factory.
 *
 * ```
 * const configureSomething = createConfigureEdges(async ({ config, vector: { to } }: OmniEdge<{ something: string }>, sdk) => {
 *   const something = await sdk.getSomethingFor(to.eid)
 *   if (something !== config.something) return []
 *
 *   return sdk.setSomething(to.eid, config.something)
 * })
 * ```
 *
 * @template TOmniGraph
 * @template TOmniSDK
 * @param {CreateTransactionsFromOmniEdges<TOmniGraph, TOmniSDK>} createTransactions
 * @returns {Configurator<TOmniGraph, TOmniSDK>}
 */
export const createConfigureEdges =
    <TOmniGraph extends OmniGraph = OmniGraph, TOmniSDK = IOmniSDK>(
        createTransactions: CreateTransactionsFromOmniEdges<TOmniGraph, TOmniSDK>
    ): Configurator<TOmniGraph, TOmniSDK> =>
    async (graph, createSdk) =>
        flattenTransactions(
            await Promise.all(
                graph.connections.map(async (edge) => {
                    const sdk = await createSdk(edge.vector.from)

                    return await createTransactions(edge, sdk, graph, createSdk)
                })
            )
        )
