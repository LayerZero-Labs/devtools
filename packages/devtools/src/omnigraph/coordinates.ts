import { EndpointId, endpointIdToStage } from '@layerzerolabs/lz-definitions'
import { OmniVector, OmniPoint, OmniNode, WithEid } from './types'

/**
 * Compares two points by value
 *
 * @param a `OmniPoint`
 * @param b `OmniPoint`
 *
 * @returns `true` if the vector point to the same point in omniverse
 */
export const arePointsEqual = (a: OmniPoint, b: OmniPoint): boolean => a.address === b.address && a.eid === b.eid

/**
 * Checks if two points are on the same endpoint
 *
 * @param a `OmniPoint`
 * @param b `OmniPoint`
 *
 * @returns `true` if the vector point to the same point in omniverse
 */
export const areSameEndpoint = (a: OmniPoint, b: OmniPoint): boolean => a.eid === b.eid

/**
 * Compares two vectors by value
 *
 * @param a `OmniVector`
 * @param b `OmniVector`
 *
 * @returns `true` if the vector point from and to the same point in omniverse
 */
export const areVectorsEqual = (a: OmniVector, b: OmniVector): boolean =>
    arePointsEqual(a.from, b.from) && arePointsEqual(a.to, b.to)

/**
 * Checks that a vector is _possible_ - i.e. connects two endpoints
 * that can be connected in reality.
 *
 * @param vector `OmniVector`
 *
 * @returns `true` if two points of the vector can be connected in reality
 */
export const isVectorPossible = ({ from, to }: OmniVector): boolean =>
    endpointIdToStage(from.eid) === endpointIdToStage(to.eid)

/**
 * Serializes a point. Useful for when points need to be used in Map
 * where we cannot adjust the default behavior of using a reference equality
 *
 * @param point `OmniPoint`
 *
 * @returns `string`
 */
export const serializePoint = ({ address, eid }: OmniPoint): string => `${eid}|${address}`

/**
 * Serializes a vector. Useful for when vectors need to be used in Map
 * where we cannot adjust the default behavior of using a reference equality
 *
 * @param point `OmniVector`
 *
 * @returns `string`
 */
export const serializeVector = ({ from, to }: OmniVector): string => `${serializePoint(from)} â†’ ${serializePoint(to)}`

/**
 * Helper function to quickly convert a pair of nodes to a vector
 *
 * @param a `OmniNode`
 * @param b `OmniNode`
 * @returns `OmniVector`
 */
export const vectorFromNodes = (a: OmniNode, b: OmniNode): OmniVector => ({ from: a.point, to: b.point })

/**
 * Helper function to add an `eid` to arbitrary values, converting them to `WithEid`
 *
 * This is useful to reduce repetition when e.g. creating multiple `OmniPoint` instances on the same network:
 *
 * ```
 * const onMainnet = withEid(EndpointId.ETHEREUM_V2_MAINNET)
 *
 * const onePoint = onMainnet({ address: '0x0' })
 * const anotherPoint = onMainnet({ address: '0x1' })
 * ```
 */
export const withEid =
    (eid: EndpointId) =>
    <T>(value: T): WithEid<T> => ({ ...value, eid })
