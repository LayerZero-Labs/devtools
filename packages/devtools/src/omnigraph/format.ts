import { EndpointId } from '@layerzerolabs/lz-definitions'
import type { OmniPoint, OmniVector } from './types'

export const formatEid = (eid: EndpointId): string => EndpointId[eid] ?? `Unknown EndpointId (${eid})`

export const formatOmniPoint = ({ eid, address, contractName }: OmniPoint): string =>
    `[${address}${contractName ? ` (${contractName})` : ``} @ ${formatEid(eid)}]`

export const formatOmniVector = ({ from, to }: OmniVector): string =>
    `${formatOmniPoint(from)} â†’ ${formatOmniPoint(to)}`
