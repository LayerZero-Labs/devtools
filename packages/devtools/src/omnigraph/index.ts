export * from './builder'
export * from './config'
export * from './coordinates'
export * from './map'
export * from './format'
export * from './schema'
export * from './types'
