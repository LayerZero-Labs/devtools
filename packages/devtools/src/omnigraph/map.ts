import { AbstractMap } from '@/common/map'
import { serializePoint, serializeVector } from './coordinates'
import type { OmniPoint, OmniVector } from './types'

export class OmniPointMap<V, K extends OmniPoint = OmniPoint> extends AbstractMap<K, V> {
    [Symbol.toStringTag] = 'OmniPointMap'

    protected override hash(point: K) {
        return serializePoint(point)
    }
}

export class OmniVectorMap<V, K extends OmniVector = OmniVector> extends AbstractMap<K, V> {
    [Symbol.toStringTag] = 'OmniVectorMap'

    protected override hash(vector: K) {
        return serializeVector(vector)
    }
}
