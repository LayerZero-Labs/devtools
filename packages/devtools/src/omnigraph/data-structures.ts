import { HashMap } from '@/common/map'
import { serializePoint, serializeVector } from './coordinates'
import type { OmniPoint, OmniVector } from './types'

export class OmniPointMap<V, P extends OmniPoint = OmniPoint> extends HashMap<P, V> {
    [Symbol.toStringTag] = 'OmniPointMap'

    protected override serializeKey(point: P): string {
        return serializePoint(point)
    }
}

export class OmniVectorMap<V, P extends OmniVector = OmniVector> extends HashMap<P, V> {
    [Symbol.toStringTag] = 'OmniVectorMap'

    protected override serializeKey(point: P): string {
        return serializeVector(point)
    }
}
