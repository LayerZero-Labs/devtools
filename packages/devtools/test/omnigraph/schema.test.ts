import fc from 'fast-check'
import { pointArbitrary, vectorArbitrary } from '@layerzerolabs/test-devtools'
import { UIntBigIntSchema, UIntNumberSchema, createOmniEdgeSchema, createOmniNodeSchema } from '@/omnigraph/schema'
import { ZodError, z } from 'zod'

describe('omnigraph/schema', () => {
    interface TestCase {
        configSchema: z.ZodSchema
        good: fc.Arbitrary<unknown>
        bad: fc.Arbitrary<unknown>
    }

    const TEST_CASES: TestCase[] = [
        {
            configSchema: z.string(),
            good: fc.string(),
            bad: fc.anything().filter((value) => typeof value !== 'string'),
        },
        {
            configSchema: z.object({ a: z.number().nonnegative() }),
            good: fc.record({ a: fc.integer({ min: 0 }) }),
            bad: fc.oneof(fc.constant({ a: -1 }), fc.string(), fc.integer(), fc.date()),
        },
    ]

    describe('createOmniNodeSchema', () => {
        describe.each(TEST_CASES)(`schema`, ({ configSchema, good, bad }) => {
            const schema = createOmniNodeSchema(configSchema)

            it('should parse successfully', () => {
                fc.assert(
                    fc.property(pointArbitrary, good, (point, config) => {
                        expect(schema.safeParse({ point, config }).success).toBeTruthy()
                    })
                )
            })

            it('should not parse', () => {
                fc.assert(
                    fc.property(pointArbitrary, bad, (point, config) => {
                        expect(schema.safeParse({ point, config }).success).toBeFalsy()
                    })
                )
            })
        })
    })

    describe('createOmniEdgeSchema', () => {
        describe.each(TEST_CASES)(`schema`, ({ configSchema, good, bad }) => {
            const schema = createOmniEdgeSchema(configSchema)

            it('should parse successfully', () => {
                fc.assert(
                    fc.property(vectorArbitrary, good, (vector, config) => {
                        expect(schema.safeParse({ vector, config }).success).toBeTruthy()
                    })
                )
            })

            it('should not parse', () => {
                fc.assert(
                    fc.property(vectorArbitrary, bad, (vector, config) => {
                        expect(schema.safeParse({ vector, config }).success).toBeFalsy()
                    })
                )
            })
        })
    })

    describe('UIntBigIntSchema', () => {
        it('should work for non-negative bigints', () => {
            fc.assert(
                fc.property(fc.bigInt({ min: BigInt(0) }), (value) => {
                    expect(UIntBigIntSchema.parse(value)).toBe(value)
                })
            )
        })

        it('should not work for negative bigints', () => {
            fc.assert(
                fc.property(fc.bigInt({ max: BigInt(-1) }), (value) => {
                    expect(() => UIntBigIntSchema.parse(value)).toThrow()
                })
            )
        })

        it('should work for non-negative bigint strings', () => {
            fc.assert(
                fc.property(fc.bigInt({ min: BigInt(0) }), (value) => {
                    expect(UIntBigIntSchema.parse(String(value))).toBe(value)
                })
            )
        })

        it('should not work for negative bigint strings', () => {
            fc.assert(
                fc.property(fc.bigInt({ max: BigInt(-1) }), (value) => {
                    expect(() => UIntBigIntSchema.parse(String(value))).toThrow()
                })
            )
        })

        it('should work for non-negative integers', () => {
            fc.assert(
                fc.property(fc.integer({ min: 0 }), (value) => {
                    expect(UIntBigIntSchema.parse(value)).toBe(BigInt(value))
                })
            )
        })

        it('should not work for negative integers', () => {
            fc.assert(
                fc.property(fc.integer({ max: -1 }), (value) => {
                    expect(() => UIntBigIntSchema.parse(value)).toThrow()
                })
            )
        })

        it('should work for non-negative integer strings', () => {
            fc.assert(
                fc.property(fc.integer({ min: 0 }), (value) => {
                    expect(UIntBigIntSchema.parse(String(value))).toBe(BigInt(value))
                })
            )
        })

        it('should not work for negative integer strings', () => {
            fc.assert(
                fc.property(fc.integer({ max: -1 }), (value) => {
                    expect(() => UIntBigIntSchema.parse(String(value))).toThrow()
                })
            )
        })

        it('should throw a ZodError for invalid values', () => {
            fc.assert(
                fc.property(fc.anything(), (value) => {
                    let isValid = false

                    try {
                        isValid = UIntBigIntSchema.safeParse(value).success
                    } catch {
                        // This catch block is designed to catch errors caused by messed up standard methods
                        // like toString or toValue that fast check throws at us
                    }

                    fc.pre(!isValid)

                    // Here we expect that whatever we got, we'll not receive a SyntaxError
                    // (coming from a BigInt() constructor) but a ZodError
                    expect(() => UIntBigIntSchema.parse(value)).toThrow(ZodError)
                })
            )
        })

        it('should mention the object path for undefined values', () => {
            const ObjectSchema = z.object({ value: UIntBigIntSchema })

            expect(() => ObjectSchema.parse({ value: undefined })).toThrowErrorMatchingSnapshot()
        })
    })

    describe('UIntNumberSchema', () => {
        it('should work for non-negative integers', () => {
            fc.assert(
                fc.property(fc.integer({ min: 0 }), (value) => {
                    expect(UIntNumberSchema.parse(value)).toBe(value)
                })
            )
        })

        it('should not work for negative integers', () => {
            fc.assert(
                fc.property(fc.integer({ max: -1 }), (value) => {
                    expect(() => UIntNumberSchema.parse(value)).toThrow()
                })
            )
        })

        it('should work for non-negative integer strings', () => {
            fc.assert(
                fc.property(fc.integer({ min: 0 }), (value) => {
                    expect(UIntNumberSchema.parse(String(value))).toBe(value)
                })
            )
        })

        it('should not work for negative integer strings', () => {
            fc.assert(
                fc.property(fc.integer({ max: -1 }), (value) => {
                    expect(() => UIntNumberSchema.parse(String(value))).toThrow()
                })
            )
        })

        it('should throw a ZodError for invalid values', () => {
            fc.assert(
                fc.property(fc.anything(), (value) => {
                    let isValid = false

                    try {
                        isValid = UIntNumberSchema.safeParse(value).success
                    } catch {
                        // This catch block is designed to catch errors caused by messed up standard methods
                        // like toString or toValue that fast check throws at us
                    }

                    fc.pre(!isValid)

                    // Here we expect that whatever we got, we'll not receive a SyntaxError
                    // (coming from a BigInt() constructor) but a ZodError
                    expect(() => UIntNumberSchema.parse(value)).toThrow(ZodError)
                })
            )
        })
    })
})
