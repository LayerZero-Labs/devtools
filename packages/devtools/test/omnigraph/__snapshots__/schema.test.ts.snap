// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`omnigraph/schema UIntBigIntSchema should mention the object path for undefined values 1`] = `
"[
  {
    "code": "custom",
    "message": "Invalid BigInt-like value",
    "path": [
      "value"
    ]
  }
]"
`;
